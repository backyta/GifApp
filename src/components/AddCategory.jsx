
import PropTypes from 'prop-types';
import { useState } from "react";

export const AddCategory = ({onNewCategory}) => {

    const[inputValue, setInputValue]= useState('');

    const onInputChange = ({target}) =>{
        setInputValue(target.value);
    }

    const onSubmit = (event) =>{
        event.preventDefault();
        if ( inputValue.trim().length <= 1 ) return; //* mas de un caracter

        // console.log( inputValue );
        // console.log(setCategories);
        // setCategories( categories =>  [inputValue, ...categories] ); 
        //* categorias en el mestado actual sin necesidad de enviar un porps de categories

        onNewCategory( inputValue.trim() );
        setInputValue('');
    }

    return ( 

        <form onSubmit={ onSubmit }>
            <input 
                type="text"
                placeholder="Buscar Gifs"
                value={ inputValue }
                onChange={ onInputChange }//* recive el evento de onchange como parametro
            />
        </form>


     );
}

AddCategory.propTypes = {
    onNewCategory: PropTypes.func.isRequired
}