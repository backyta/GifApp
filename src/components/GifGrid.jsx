import PropTypes from 'prop-types';

import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ({category}) => {

    const { images, isLoading } = useFetchGifs( category );
    // console.log({ images, isLoading });

    return ( 
        <>
            <h3>{ category }</h3>
            {
                //* Mostrar mensaje de carga
                isLoading && ( <h2>Cargando....</h2> ) //* remplaza al ternario, otra opcion no usa null :
                
            }
            

            <div className='card-grid'>
                { 
                    images.map( ( image ) => (
                        <GifItem 
                            key={image.id} 
                            { ...image } //* esparcir el objeto para que tenga todas sus propiedades
                        />
                    ))
                }  
            </div>

        </>
     );
}
    
GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

    
//* Nunca se debe colocar la ejecucion de una funcion deirecto en un componente, porquie cada vez que 
//* el functional component se vuelva a renderizar va volver a ejecutar la funcion, por consecuencia 
//* vuelve a ejecutar la peticion http.

//* Si ya recibimos la categoria solo se ejecuta una vez esa funcion, la peticion no se volvera a asignar
//* esapcio en memoria por eso se coloca fuera. No importa si se redibuja, la peticion no se volvera a lanzar

//? USE EFECT
//* Es un hook de react que sirve para disparar efectos secundarios, se entiendo como algun proceso que queremos
//* ejecutar cuando algo suceda.

//* Recibe 2 parametros, el efecto que es un callback ques una funcion, y luego una lista de dependencias
//* que son las condiciones por las cuales queremos volver a ejecutar el callback.

//* Si se deja las dependencias vacias significa que el hook solo se va disparar la priemra vez que se crea
//* y se construye el componente.

//? UseState y UseEfect
//* useState mantiene/guarda/recuerda los valores de tu Componente, esto es en lo que React se "especializa" o 
//* el principal trabajo que hace.

//? Use Efect
//* useEffect (efectos secundarios) es como un puente entre tu componente y el mundo exterior, como:
    //* API en un servidor remoto: Donde quieres recibir informacion y mostrarlo en tu Componente   
    //* (actualizando su estado).
    //* Eventos del mouse en la pantalla: Puedes escuchar estos eventos y actualizar tu estado del Componente.