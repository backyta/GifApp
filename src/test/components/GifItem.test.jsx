/* eslint-disable no-undef */
import { render, screen } from "@testing-library/react";
import { GifItem } from "../../components/GifItem";

describe('Pruebas en Component GifItem', () => {

    const props = {
        title: 'Ttitulo prueba',
        url: 'http://localhost/url-prueba'
    }

    test('Debe de hacer match con el snapshot', () => {
        const { container } = render( <GifItem { ...props } /> )
        expect( container ).toMatchSnapshot();
        
    });

    test('Debe de mostrar lka imagen con la URL y el ALT indicado', () => {
        render( <GifItem { ...props } /> );

        // screen.debug();
        // expect( screen.getByRole('img').src ).toBe( props.url ) 
        //* manera resumida y elegante, destructuring
        const { src, alt } = screen.getByRole('img');
        expect( src ).toBe( props.url );
        expect( alt ).toBe( props.title );

    });

    test('Debe de mostrar el titulo en el componente', () => {

        render( <GifItem { ...props } /> )
        expect( screen.getByText( title )).toBeTruthy(); 
        
        //* evalua un texto no una propiedad de etiqueta que llava el mismo titulo
    });
    
});

//* El snapshot no me asegura de que la imagen este correctemente colocado o el titulo.
//* Snapshot asegura que al momento de ejecutar las pruebas toda las piezas del HTML sintacticamenter
//* esten igual como yo tome la fotografia inicialmente.
