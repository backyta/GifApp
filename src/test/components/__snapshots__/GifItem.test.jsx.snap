// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Component GifItem Debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Ttitulo prueba"
      src="http://localhost/url-prueba"
    />
    <p>
      Ttitulo prueba
    </p>
  </div>
</div>
`;
